File: 2017/lua/day08.lua
require("stuff")

ops = { ["<"] = function(r, n) return regs[r] < n end,
        [">"] = function(r, n) return regs[r] > n end,
        ["<="] = function(r, n) return regs[r] <= n end,
        [">="] = function(r, n) return regs[r] >= n end,
        ["=="] = function(r, n) return regs[r] == n end,
        ["!="] = function(r, n) return regs[r] ~= n end }

regs, max = {}, 0
for line in io.lines("08.txt") do
    local r1, sign, n1, r2, op, n2 = string.match(line, "(%w+) (%w+) (%-?%d+) if (%w+) ([!<>=]=?) (%-?%d+)")
    if sign == "dec" then n1 = -n1 end
    
    if not regs[r1] then regs[r1] = 0 end
    if not regs[r2] then regs[r2] = 0 end
    if ops[op](r2, tonumber(n2)) then regs[r1] = regs[r1] + n1 end
    if regs[r1] > max then max = regs[r1] end
end

print(string.format("Part 1: %d", stuff.tablemax(regs)))
print(string.format("Part 2: %d", max))

File: 2017/kotlin/day08.kt
import java.io.File

fun op(operator: String, a: Int, b: Int): Boolean = when(operator) {
        "<=" -> a <= b
        ">=" -> a >= b
        "<" -> a < b
        ">" -> a > b
        "==" -> a == b
        else -> a != b
    }

fun main() {
    val registers = mutableMapOf<String, Int>()
    var maxvalue = 0
    for (line in File("08.txt").readLines()) {
        val (instruction, condition) = Regex("""(\w+) (\w+|[<>!=]=?) (-?\d+)""").findAll(line).map { it.value }.toList()
        val check = with(condition.split(" ")) { op(this[1], (registers[this[0]] ?: 0), this[2].toInt()) }
        if (check) with(instruction.split(" ")) {
            registers[this[0]] = (registers[this[0]] ?: 0) + (if (this[1] == "dec") -1 else 1) * this[2].toInt()
            if (registers[this[0]]!! > maxvalue) maxvalue = registers[this[0]]!!
        }
    }
    
    println("Part 1 answer: ${registers.values.max()}")
    println("Part 2 answer: $maxvalue")
}